### PR TITLE
Function to compute endpoints

### PR DESCRIPTION
**Pull request by chiraganand**
Return a vector of index values for last observation in `ts` for each period given by `on`, ending with the last values of `index(ts)`. The values are picked up every `k` instance of period. The last observation of each unique group (based on `on`) is returned.

## Supported interfaces
```julia
endpoints(timestamps::AbstractVector{T}, on::V) where {T<:Union{Date, DateTime, Time},
                                                       V<:Union{
                                                           Year,
                                                           Quarter,
                                                           Month,
                                                           Week,
                                                           Day,
                                                           Hour,
                                                           Minute,
                                                           Second,
                                                           Millisecond,
                                                           Microsecond,
                                                           Nanosecond
                                                       }}
endpoints(ts::TS, on::T) where {T<:Dates.Period}
endpoints(ts::TS, on::Symbol, k::Int=1)
endpoints(ts::TS, on::String, k::Int=1)
endpoints(ts::TS, on::Function, k::Int=1)
endpoints(values::AbstractVector, on::Function, k::Int=1)
```

The primary interface works on all time type vectors, namely, `Date`, `DateTime`, and `Time` and on all sub-types of `Dates.Period`. Rest of the methods except the ones accepting `::Function` type for `on` call the primary method with the relevant arguments.

`endpoints(values::AbstractVector, on::Function, k::Int=1)` allows one to pass a function which creates unique groups which will then be used for getting the last observation from.

## Examples
```julia
julia> ts = TS(random(length(dates)), dates)
(731 x 1) TS with Date Index

 Index       x1
 Date        Float64
───────────────────────
 2017-01-01  0.768448
 2017-01-02  0.940515
 2017-01-03  0.673959
 2017-01-04  0.395453
 2017-01-05  0.313244
 2017-01-06  0.662555
 2017-01-07  0.586022
 2017-01-08  0.0521332
 2017-01-09  0.26864
 2017-01-10  0.108871
     ⋮           ⋮
 2018-12-24  0.812797
 2018-12-25  0.158056
 2018-12-26  0.269285
 2018-12-27  0.15065
 2018-12-28  0.916177
 2018-12-29  0.278016
 2018-12-30  0.617211
 2018-12-31  0.67549
 2019-01-01  0.910285
       712 rows omitted

julia> ts[endpoints(ts, Month(1))]
(25 x 1) TS with Date Index

 Index       x1
 Date        Float64
───────────────────────
 2017-01-31  0.48
 2017-02-28  0.458476
 2017-03-31  0.274441
 2017-04-30  0.413966
 2017-05-31  0.734931
 2017-06-30  0.257159
 2017-07-31  0.415851
 2017-08-31  0.0377973
 2017-09-30  0.934059
 2017-10-31  0.413175
 2017-11-30  0.557009
 2017-12-31  0.346659
 2018-01-31  0.174777
 2018-02-28  0.432223
 2018-03-31  0.835142
 2018-04-30  0.945539
 2018-05-31  0.0635483
 2018-06-30  0.589922
 2018-07-31  0.285088
 2018-08-31  0.912558
 2018-09-30  0.238931
 2018-10-31  0.49775
 2018-11-30  0.830232
 2018-12-31  0.67549
 2019-01-01  0.910285

julia> ts[endpoints(ts, Month(2))]
(12 x 1) TS with Date Index

 Index       x1
 Date        Float64
───────────────────────
 2017-02-28  0.458476
 2017-04-30  0.413966
 2017-06-30  0.257159
 2017-08-31  0.0377973
 2017-10-31  0.413175
 2017-12-31  0.346659
 2018-02-28  0.432223
 2018-04-30  0.945539
 2018-06-30  0.589922
 2018-08-31  0.912558
 2018-10-31  0.49775
 2018-12-31  0.67549
 2019-01-01  0.910285
```

Refs #51.

### CODE REVIEW

File: Project.toml [hunk 1]
```diff
@@ -6,6 +6,7 @@ version = "0.1.0"
 [deps]
 DataFrames = "a93c6f00-e57d-5684-b7b6-d8193f3e46c0"
 Dates = "ade2ca70-3891-5945-98fb-dc099432e06a"
+Random = "9a3f8284-a2c9-5f02-9a11-845980a1fd5c"
```
Comment (ValentinKaisermayer) [inline Project.toml:9]: Why is `Random` a dependency of the package?

Comment (chiraganand) [reply Project.toml:9]: It is being used for reproducible test cases as well as examples in the documentation.

**Comment by chiraganand** [comment]
Done with this PR from my end. Ready for review. @ValentinKaisermayer would you please?

**Review by ayushpatnaikgit — APPROVED** [review]

**Comment by chiraganand** [comment]
Merging this now so that rest of the development can proceed. Any further issues can be resolved later.

**Pull request merged**